operators/advanced_network/advanced_network/managers/dpdk/adv_network_dpdk_mgr.cpp

```diff
@@ -1538,150 +1532,118 @@ int DpdkMgr::rx_core_worker(void* arg) {
-          pkt_info[p].flow_id = mbuf_arr[to_copy + p]->hash.fdir.hi;
-        } else {
-          pkt_info[p].flow_id = 0;
+      for (int seg = 0; seg < tparams->num_segs; seg++) {
+        if (rte_mempool_get(tparams->burst_pool, reinterpret_cast<void**>(&burst->pkts[seg])) < 0) {
+          HOLOSCAN_LOG_ERROR(
+              "Processing function falling behind. No free RX bursts!");
+          continue;
         }
       }
 
-      if (tparams->num_segs > 1) {  // Extra work when buffers are scattered
-        for (int p = 0; p < nb_rx; p++) {
-          struct rte_mbuf* mbuf = mbuf_arr[to_copy + p];
-          for (int seg = 1; seg < tparams->num_segs; seg++) {
-            mbuf = mbuf->next;
-            burst->pkts[seg][p] = mbuf;
-          }
-        }
-
-        cur_pkt_in_batch += nb_rx;
+      if (rte_mempool_get(
+            tparams->flowid_pool, reinterpret_cast<void**>(&burst->pkt_extra_info)) < 0) {
+        HOLOSCAN_LOG_ERROR("Processing function falling behind. No free CPU buffers for packets!");
+        continue;
       }
 
-      nb_rx = 0;
-    } else {
-      burst->hdr.hdr.num_pkts = 0;
+      pkt_info = reinterpret_cast<ExtraRxPacketInfo*>(burst->pkt_extra_info);
     }
 
-    // DPDK on some ARM platforms requires that you always pass nb_pkts as a number divisible
-    // by 4. If you pass something other than that, you get undefined results and will end up
-    // running out of buffers.
-    do {
-      int burst_size = std::min((uint32_t)DEFAULT_NUM_RX_BURST,
-                                (uint32_t)(tparams->batch_size - burst->hdr.hdr.num_pkts));
-
-      nb_rx = rte_eth_rx_burst(tparams->port,
-                               tparams->queue,
-                               reinterpret_cast<rte_mbuf**>(&mbuf_arr[0]),
-                               DEFAULT_NUM_RX_BURST);
+    // Check if we need to get more packets
+    if (nb_rx == 0) {
+      cur_pkt_in_batch = 0;
+      nb_rx = rte_eth_rx_burst(tparams->port, tparams->queue,
+                          reinterpret_cast<rte_mbuf**>(&mbuf_arr[0]), DEFAULT_NUM_RX_BURST);
 
       if (nb_rx == 0) {
         if (burst->hdr.hdr.num_pkts > 0 && timeout_cycles > 0) {
           const auto cur_cycles = rte_get_tsc_cycles();
 
           // We hit our timeout. Send the partial batch immediately
           if ((cur_cycles - last_cycles) > timeout_cycles) {
-            cur_pkt_in_batch = 0;
             rte_ring_enqueue(tparams->ring, reinterpret_cast<void*>(burst));
             last_cycles = cur_cycles;
-            break;
+            burst = nullptr;
           }
         }
 
         continue;
       }
+    }
 
-      to_copy = std::min(nb_rx, (int)(tparams->batch_size - burst->hdr.hdr.num_pkts));
-      memcpy(&burst->pkts[0][burst->hdr.hdr.num_pkts], &mbuf_arr, sizeof(rte_mbuf*) * to_copy);
+    // At this point we have some packets to copy either from a new batch or an existing one. Check
+    // if we are finishing a batch or copying all packets that came in
```

Comment:
Naive question as I don't get most of this code, but to help make sure I learn some of it, how come we're doing a copy?

Reply:
this is copying the burst headers from the local array into the ANO burst array. No data is copied here